File: quicksort.py
from typing import List


def partition (input_array: List[int], low: int, high: int):
    pivot = input_array[low] #-1
    leftwall = low #0

    for i in range(low + 1, high ): #1-3
        if (input_array[i] < pivot):
            temp = input_array[i]
            input_array[i] = input_array[leftwall]
            input_array[leftwall] = temp
            leftwall = leftwall + 1

    temp2 = pivot
    pivot = input_array[leftwall]
    input_array[leftwall] = temp2

    return leftwall
def quicksort(input_array: List[int], low: int, high: int):
    #[-1, 3, 2, 1]
    if (low < high):
        pivot_location = partition(input_array, low, high)
        quicksort(input_array, low, pivot_location)
        quicksort(input_array, pivot_location + 1, high)

File: quickSort.java
import java.util.Arrays;
import static java.lang.Math.abs;
import java.util.Random;

public class quickSort {

    public static int partition(int[] array, int left, int right) {
        int i = left;
        int j = right - 1;
        int pivot = array[right];
        int temp = 0;

        while (i < j) {
            while (i < j && array[i] <= pivot) {
                i += 1;
            }
            while (j > i && array[j] > pivot) {
                j -= 1;
            }
            if (array[i] > array[j]) {
                temp = array[i];
                array[i] = array[j];
                array[j] = temp;
            }
        }
        if (array[i] > pivot) {
            temp = array[i];
            array[i] = array[right];
            array[right] = temp;
        } else {
            i = right;
        }
        return i;
    }

    public static void quickSort(int[] array, int left, int right) {
        if (left < right) {
            int pivotLocation = partition(array, left, right);
            quickSort(array, left, pivotLocation - 1);
            quickSort(array, pivotLocation + 1, right);
        }
    }

    public static void main(String[] args) {

        int arrayStart = 0;
        int[] array_sizes =  new int[]{10, 100, 500, 1000, 1500, 2000, 2500, 3000, 5000, 7500, 10000};
        for (int i = 0; i < array_sizes.length; i++){
            Random rd = new Random();
            int[] arr = new int[array_sizes[i]];
            for (int j = 0; j < arr.length; j++) {
                arr[i] = rd.nextInt(); // storing random integers in an array
            }
            long timeStart = System.nanoTime();
            quickSort(arr, arrayStart, arr.length-1);
            long timeEnd = System.nanoTime();
            System.out.println("Verlaufszeit der Schleife: " + (timeEnd - timeStart) + " Millisek.");
        }
    }
}